contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1146 @@
+            maxHeldToken = TokenInteract.balanceOf(HELD_TOKEN, address(this));
+        }
+
+        for (uint256 i = 0; i < buckets.length; i++) {
+            uint256 bucket = buckets[i];
+
+            // prevent withdrawing from the current bucket if it is also the critical bucket
+            if (bucket != 0 && bucket == lockedBucket) {
+                continue;
+            }
+
+            (uint256 owedTokenForBucket, uint256 heldTokenForBucket) = withdrawSingleBucket(
+                bucket,
+                maxWeights[i],
+                maxHeldToken
+            );
+
+            totalOwedToken = totalOwedToken.add(owedTokenForBucket);
+            totalHeldToken = totalHeldToken.add(heldTokenForBucket);
+        }
+
+        // Transfer share of owedToken
+        TokenInteract.transfer(OWED_TOKEN, beneficiary, totalOwedToken);
+        TokenInteract.transfer(HELD_TOKEN, beneficiary, totalHeldToken);
+
+        return (totalOwedToken, totalHeldToken);
+    }
+
+    // ============ Helper Functions ============
+
+    /**
+     * Recalculates the Outstanding Principal and Available Amount for the buckets. Only changes the
+     * state if part of the position has been closed since the last position increase.
+     */
+    function rebalanceBucketsInternal()
+        internal
+    {
+        // if force-closed, don't update the outstanding principal values; they are needed to repay
+        // lenders with heldToken
+        if (wasForceClosed) {
+            return;
+        }
+
+        uint256 marginPrincipal = getPositionPrincipal();
+
+        accountForClose(principalTotal.sub(marginPrincipal));
+
+        assert(principalTotal == marginPrincipal);
+    }
+
+    /**
+     * Updates the state variables at any time. Only does anything after the position has been
+     * closed or partially-closed since the last time this function was called.
+     *
+     * - Increases the available amount in the highest buckets with outstanding principal
+     * - Decreases the principal amount in those buckets
+     *
+     * @param  principalRemoved  Amount of principal closed since the last update
+     */
+    function accountForClose(
+        uint256 principalRemoved
+    )
+        private
+    {
+        if (principalRemoved == 0) {
+            return;
+        }
+
+        uint256 newRepaidAmount = Margin(DYDX_MARGIN).getTotalOwedTokenRepaidToLender(POSITION_ID);
+        assert(newRepaidAmount.sub(cachedRepaidAmount) >= principalRemoved);
+
+        uint256 principalToSub = principalRemoved;
+        uint256 availableToAdd = newRepaidAmount.sub(cachedRepaidAmount);
+        uint256 criticalBucketTemp = criticalBucket;
+
+        // loop over buckets in reverse order starting with the critical bucket
+        for (
+            uint256 bucket = criticalBucketTemp;
+            principalToSub > 0;
+            bucket--
+        ) {
+            assert(bucket <= criticalBucketTemp); // no underflow on bucket
```

Comment:
Why don't you just use safe meth on the `bucket--`? This seems like a weird way to check for underflow

Reply:
Also - I forget if I asked this before, but this should never underflow right? As we always should have accounted for enough principal? So it should indeed be an assert?